File: Local Labor Market Data/Model/Report/OESReport.swift
//
//  OESReport.swift
//  Local Labor Market Data
//
//  Created by Nidhi Chawla on 7/31/18.
//  Copyright © 2018 Department of Labor. All rights reserved.
//

import Foundation

class OESReport {
    enum DataTypeCode: String {
        case employment = "01"
        case employmentPercentRelativeStdError = "02"
        case hourlyMeanWage = "03"
        case annualMeanWage = "04"
        case wagePercentRelativeStdError = "05"
        case hourly10PercentileWage = "06"
        case hourly25PercentileWage = "07"
        case hourlyMedianWage = "08"
        case hourly75PercentileWage = "09"
        case hourly90PercentileWage = "10"
        case annual10PercentileWage = "11"
        case annual25PercentileWage = "12"
        case annualMedianWage = "13"
        case annual75PercentileWage = "14"
        case annual90PercentileWage = "15"
        case employmentPer1000Jobs = "16"
        case locationQuotient = "17"
    }
    
    class func getSeriesId(forArea area: Area, occupationCode: String = "000000",
                           dataTypeCode: DataTypeCode,
                           adjustment: SeasonalAdjustment) -> SeriesId? {

        guard let code = area.code else { return nil}
        
        let areaType: String
        let areaCode: String
        if area is Metro {
            areaType = "M"
            areaCode = code.leftPadding(toLength: 7, withPad: "0")
        }
        else if area is State {
            areaType = "S"
            areaCode = code.rightPadding(toLength: 7, withPad: "0")
        }
        else {
            areaType = "N"
            areaCode = code.rightPadding(toLength: 7, withPad: "0")
        }
        
        let industryCode = "000000"
        let seriesId = "OE" + adjustment.rawValue  + areaType + areaCode + industryCode + occupationCode + dataTypeCode.rawValue
        return seriesId
    }
}

File: app/src/main/java/blsapp/dol/gov/blslocaldata/model/reports/OESReport.kt
package blsapp.dol.gov.blslocaldata.model.reports

import blsapp.dol.gov.blslocaldata.db.entity.AreaEntity
import blsapp.dol.gov.blslocaldata.db.entity.MetroEntity
import blsapp.dol.gov.blslocaldata.db.entity.StateEntity

class OESReport {
    enum class DataTypeCode(val code: String) {
        EMPLOYMENT("01"),
        EMPLOYMENTPERCENTRELATIVESTDERROR("02"),
        HOURLYMEANWAGE("03"),
        ANNUALMEANWAGE("04"),
        WAGEPERCENTRELATIVESTDERROR("05"),
        HOURLY10PERCENTILEWAGE("06"),
        HOURLY25PERCENTILEWAGE("07"),
        HOURLYMEDIANWAGE("08"),
        HOURLY75PERCENTILEWAGE("09"),
        HOURLY90PERCENTILEWAGE("10"),
        ANNUAL10PERCENTILEWAGE("11"),
        ANNUAL25PERCENTILEWAGE("12"),
        ANNUALMEDIANWAGE("13"),
        ANNUAL75PERCENTILEWAGE("14"),
        ANNUAL90PERCENTILEWAGE("15"),
        EMPLOYMENTPER1000JOBS("16"),
        LOCATIONQUOTIENT("17")
    }

    companion object {
        fun getSeriesId(area: AreaEntity, occupationCode: String = "000000",
                        dataTypeCode: DataTypeCode): SeriesId {

            val code = area.code
            val areaType: String
            val areaCode: String

            when(area) {
                is MetroEntity -> {
                    areaType = "M"
                    areaCode = code.padStart(7, '0')
                }
                is StateEntity -> {
                    areaType = "S"
                    areaCode = code.padEnd(7, '0')
                }
                else -> {
                    areaType = "N"
                    areaCode = code.padEnd(7, '0')
                }
            }

            val industryCode = "000000"
            // OES data are annual and therefore do not need to be seasonally adjusted.
            // The same data should appear whether the seasonally adjusted toggle is turned on or off
            // let seriesId = "OE" + adjustment.rawValue  + areaType + areaCode + industryCode + occupationCode +
            // dataTypeCode.rawValue
            val seriesId = "OE" + SeasonalAdjustment.NOT_ADJUSTED.value  + areaType +
                    areaCode + industryCode + occupationCode + dataTypeCode.code

            return seriesId
        }
    }
}